### PR TITLE
Added support for Deferred Enabled events

### PR DESCRIPTION
**Pull request by SilentsReplacement**
The signal module wouldn't work properly and just error when events were deferred. This commit fixes this all properly.

### CODE REVIEW

File: src/Util/Signal.lua [hunk 1]
```diff
@@ -144,14 +153,33 @@ end
...
 function Signal:Connect(handler)
-	local connection = Connection.new(self, self._bindable.Event:Connect(function(id)
+	self._handlers += 1
+	self._handlersLeft += 1
+	
+	local connection
+	connection = Connection.new(self, self._bindable.Event:Connect(function(id)
+		if not connection:IsConnected() then
+			-- Deferred enabled events are queued
+			warn("ye")
```
Comment (SilentsReplacement) [inline src/Util/Signal.lua:163]: Remove this warn as well,  I forgot to remove it during testing

**Comment by LucasMZReal** [comment]
Hm, you said it was completely broken? Are you sure? I can get it to work just fine on Deferred Events without this commit? At least it didn't error at all for me. What code were you using?

File: src/Util/Signal.lua [hunk 2]
```diff
@@ -163,15 +191,40 @@ function Signal:DisconnectAll()
...
+		if self._threadsLeft > 0 then
+			self._threadsCompleted.Event:Wait()
+		end
+
+		if self._handlersLeft > 0 then
+			self._handlersCompleted.Event:Wait()
+		end
+
+		self._connections = {}
+		self._args = {}
```
Comment (LucasMZReal) [inline src/Util/Signal.lua:206]: Would `table.clear` be better here?

**Comment by SilentsReplacement** [comment]
@LucasMZReal The signal module is in fact broken, it will error for you once you start using Knit and have some communication between services / controllers. Even if you don't use Knit or have some communication, it will still have a lot of edge cases to handle without this commit and and when those are left unhandled, it will be completely broken.

> Also the connection:IsConnected() part when a event is fired seems wrong to use, that should be something that should be changed on the script someone is using; and not changed on the Signal library as that would make it inconsistent to actual RBXScriptSignal behaviour;

This is not wrong to use. Bindable events will be deferred, and that means that the handler will be queued to resume at a later time under a specific invocation point. So if you fire a event and at the same, disconnect the event, the handler will still be called. This is also mentioned in the [thread](https://devforum.roblox.com/t/beta-deferred-lua-event-handling/1240569) it self:

image

File: src/Util/Signal.lua (hunk 2 above)
Comment (SilentsReplacement) [reply src/Util/Signal.lua:206]: No need, redundant.

**Pull request closed without merging**